### DISCUSSION TITLE
Strict print width

### DISCUSSION THREAD

**Discussion by RunDevelopment** [hosseinmd/prettier-plugin-jsdoc#65] [General]
Let me be the first one to start a discussion.

Right now, this plugin ignores the `printWidth` setting (explicitly specified by the user) in some situations. It does this because the end of paragraphs may be very short otherwise which isn't nice visually. See the [full explanation with examples here](https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/26#issuecomment-771842456).

The purpose of this discussion is to determine whether this plugin should be allowed to ignore the `printWidth` setting for a more visually pleasing result.

---

My opinion on the matter is simple: No, it should not ignore `printWidth`.

`printWidth` was explicitly set by the user. We can't just ignore that. 

I would also argue that paragraphs that are randomly longer than the rest look a lot worse than paragraphs ending with a single word.

<details>

(`printWidth` ignored)
![image](https://user-images.githubusercontent.com/20878432/107442860-0c155f80-6b38-11eb-807d-6b560c32567f.png)

(`printWidth` honored)
![image](https://user-images.githubusercontent.com/20878432/107442955-2e0ee200-6b38-11eb-8b47-04ae6db04b8e.png)

</details>

I would like to point out that the question of where to insert line breaks, is a problem that programs like MS Words, LibreOffice, and LaTeX renders also have to solve, but they can't print beyond the page. How do those programs solve this?

**Comment by fardad-dev** [comment]
Thank you for contribute.

**Reply by fardad-dev** [reply]
MS Office changing space between characters to justify content. I don't know it is possible for us.

**Reply by RunDevelopment** [reply]
IDE's typically use monospaced fonts, so this isn't possible.

**Comment by RunDevelopment** [comment]
I had an idea. Print width is ignored to prevent a situation where the last line of a paragraph contains just a single word or two. IMO, this isn't a problem, but how about we do this:

Instead of appending the last very short line to the one above (going beyond print width), how about we move a few words from the second last line down to the last line?

<details>
<summary>Example</summary>


(print width ignored)
```
Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo 
ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis 
dis parturient montes, nascetur ridiculus mus. Donec quam felis, amet
ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa a enim.
```

(print width honored)
```
Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo 
ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis 
dis parturient montes, nascetur ridiculus mus. Donec quam felis, amet
ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa 
a enim.
```

(new method)
```
Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo 
ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis 
dis parturient montes, nascetur ridiculus mus. Donec quam felis, amet
ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat
massa a enim.
```

</details>

How many characters/words of the second last line will be moved into the last one also has to be decided somehow.

**Reply by fardad-dev** [reply]
I like it, but that will be break single line comments

**Reply by RunDevelopment** [reply]
You mean paragraphs that are just slightly too long to fit into a single line, right? If they are too long, then the same rules will apply. The goal is to honor print width after all.

<details>
<summary>Example</summary>

(print width ignored)
```
Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo 
ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis 
dis parturient montes, nascetur ridiculus mus. 

Donec quam felis, amet ultricies nec, pellentesque eu, pretium quis, sem. Nulla.
```

(print width honored)
```
Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo 
ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis 
dis parturient montes, nascetur ridiculus mus. 

Donec quam felis, amet ultricies nec, pellentesque eu, pretium quis, sem. 
Nulla.
```

(new method)
```
Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo 
ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis 
dis parturient montes, nascetur ridiculus mus. 

Donec quam felis, amet ultricies nec, pellentesque eu, pretium quis, 
sem. Nulla.
```

</details>

**Reply by fardad-dev** [reply]
No, this example https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/26#issuecomment-771842456

**Reply by fardad-dev** [reply]
For example 

```
/** Donec quam felis, amet ultricies nec, pellentesque eu, pretium quis, sem. Nulla. */
```

**Reply by RunDevelopment** [reply]
If it's too long, it will be split into two lines.

<details>
<summary>Example</summary>

(print width ignored)
```
/** Donec quam felis, amet ultricies nec, pellentesque eu, pretium quis, sem. Nulla. */
```

(print width honored)
```
/**
 * Donec quam felis, amet ultricies nec, pellentesque eu, pretium quis, sem. 
 * Nulla.
 */
```

(new method)
```
/**
 * Donec quam felis, amet ultricies nec, pellentesque eu, pretium quis, 
 * sem. Nulla.
 */
```

</details>

It really shouldn't matter that the comment could be a single line if only print width was a little more. We can't just ignore print width when it's convenient.

**Reply by fardad-dev** [reply]
Many developers like single comment, we should keep this feature.

**Reply by fardad-dev** [reply]
A word more not so matter when it is single line

**Reply by RunDevelopment** [reply]
I personally don't like that single-line comments go beyond. But, it would be acceptable if it was the only place where printWidth is ignored.

**Comment by vincerubinetti** [comment]
@RunDevelopment in case you haven't seen, there is more discussion of this issue in https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/168 . I'm glad I'm not the only one who's complaining about this "clever" wrapping behavior. Luckily though, it seems like a user is working on a PR that will allow us to specify a wrapping strategy, and wrapping "normally" as you describe should be one of the options.
